### ISSUE TITLE
Support Kafka zstd compression

### ISSUE THREAD

**Issue by KarstenSchnitter** [influxdata/telegraf#7123]
### Description

The Kafka plugins do not support zstd compression. This is due to the used sarama version. It can be fixed by an upgrade of the dependency to versions >1.26.0. Current version is [1.26.1](https://github.com/Shopify/sarama/releases/tag/v1.26.1)

### Proposal:

Upgrade Shopify/sarama to [1.26.1](https://github.com/Shopify/sarama/releases/tag/v1.26.1)

### Current behavior:

telegraf cannot consume or produce Kafka topics with zstd compression

### Desired behavior:

telegraf can consume or produce Kafka topics with zstd compression

**Comment by sjwang90** [comment]
@KarstenSchnitter We've upgraded to sarama 1.27.1 in #8289. If you're interested in submitting a PR for Kafka zstd compression that would be great. Otherwise we'll add this to our backlog.

**Comment by KarstenSchnitter** [comment]
This is great news. I have not found the time to try that out yet. Shopify/sarama v1.27.1 should support zstd compression out of the box.

**Comment by schmorgs** [comment]
I tested this with 1.17.2 agent today, ZSTD compression set in the output pointing to Kafka running v2.5.0 but the agent would not start and shows
```E! [telegraf] Error running agent: could not initialize output kafka: kafka: invalid configuration (zstd compression requires Version >= V2_1_0_0)```

Kafka is definitely above this version so not sure where Telegraf plugin is getting the version from.

**Comment by KarstenSchnitter** [comment]
This error is generated by sarama when the Kafka consumer version is below 2.1.0. I think it is configured here to be 0.10.2.0: https://github.com/influxdata/telegraf/blob/a790529bdd9c0f097ce6e7889d8b01bdeb85a4a2/plugins/inputs/kafka_consumer/kafka_consumer.go#L180 The documentation says it should be configurable by property `version`, but I cannot find code, that would honour any configured value. If this was the case using a version above 2.1.0 should work.

**Comment by KarstenSchnitter** [comment]
I revisited the configuration of the sarama Kafka client. It parses the `version` plugin property correctly and hands it over to sarama. This property tells sarama which Kafka version to use for the client. Using zstd requires at least version 2.1.0. @schmorgs have you tried changing this property in your plugin configuration? 

```toml
[[inputs.kafka_consumer]]
  brokers = ["localhost:9092"]
  topics = ["telegraf"]
  version = "2.1.0"
  # further config
```

**Comment by schmorgs** [comment]
> I revisited the configuration of the sarama Kafka client. It parses the `version` plugin property correctly and hands it over to sarama. This property tells sarama which Kafka version to use for the client. Using zstd requires at least version 2.1.0. @schmorgs have you tried changing this property in your plugin configuration?
> 
> ```toml
> [[inputs.kafka_consumer]]
>   brokers = ["localhost:9092"]
>   topics = ["telegraf"]
>   version = "2.1.0"
>   # further config
> ```

Thanks, I assumed the Sarama code would detect the version and so didn't spot this param.

I've tested and it works fine, thanks very much for the help

**Comment by hackery** [comment]
Is there any expectation of changing the default from 0.10.2.0 or do we just have to add this override everywhere?

**Comment by KarstenSchnitter** [comment]
0.10.2.0 is the Sarama default. It is hard to decide on any other default, since the  Kafka version in use is unknown.

I have experienced similar problems with Sarama consuming zstd topics outside of Telegraf. Maybe you could try to play around with the `max_message_length` parameter which is set to 0 (unlimited) by default. I am not sure it will help though. Maybe you could also try to throttle the topic in the Kafka configuration, so that Sarama is not able to consume too much data.

**Comment by hackery** [comment]
Kafka has (ok ... [purports to have](https://www.confluent.io/blog/upgrading-apache-kafka-clients-just-got-easier/)?) excellent bidirectional client/broker version compatibility negotiation. If Telegraf selected a later version in the Sarama config, would Sarama gracefully fall back to earlier API versions if needed, or does this setting _force_ it to use the API version given?

Setting `max_message_length` is unlikely to help here; our messages are mostly short, it's the large _total_ (`count * size`) of messages in the backlog that causes the issue. Kafka broker configuration provides no options that can throttle a consumer, as far as I know. What might be useful is a way in the Telegraf plugin to set arbitrary consumer properties.

**Comment by hackery** [comment]
According to @dnwe in [sarama#1831](https://github.com/Shopify/sarama/issues/1831) the memory inflation issue was addressed in Sarama 1.28.0 - first version of Telegraf with this fix appears to be [1.20.3](https://github.com/influxdata/telegraf/blame/v1.20.3/go.mod#L35).
